Common/Util/RateGate.cs

```diff
@@ -109,31 +109,32 @@ public RateGate(int occurrences, TimeSpan timeUnit)
             _exitTimer = new Timer(ExitTimerCallback, null, TimeUnitMilliseconds, -1);
         }
 
-        // Callback for the exit timer that exits the semaphore based on exit times 
+        // Callback for the exit timer that exits the semaphore based on exit times
         // in the queue and then sets the timer for the nextexit time.
+        // Credit to Jim: http://www.jackleitch.net/2010/10/better-rate-limiting-with-dot-net/#comment-3620
+        // for providing the code below, fixing issue #3499 - https://github.com/QuantConnect/Lean/issues/3499
         private void ExitTimerCallback(object state)
         {
             // While there are exit times that are passed due still in the queue,
             // exit the semaphore and dequeue the exit time.
-            int exitTime;
-            while (_exitTimes.TryPeek(out exitTime)
-                    && unchecked(exitTime - Environment.TickCount) <= 0)
+            var exitTime = 0;
+            var exitTimeValid = _exitTimes.TryPeek(out exitTime);
+            while (exitTimeValid)
             {
+                if (unchecked(exitTime - Environment.TickCount) > 0)
+                {
+                    break;
+                }
                 _semaphore.Release();
                 _exitTimes.TryDequeue(out exitTime);
+                exitTimeValid = _exitTimes.TryPeek(out exitTime);
             }
+            // we are already holding the next item from the queue, do not peek again
+            // although this exit time may have already pass by this stmt.
+            var timeUntilNextCheck = exitTimeValid
+                ? Math.Min(TimeUnitMilliseconds, Math.Max(0, exitTime - Environment.TickCount))
```

Comment:
Not sure what precision are we interested in having here but if `exitTime - Environment.TickCount` is sudenly `<= 0` we are going to wait `TimeUnitMilliseconds` before releasing the waiting consumers, or if `exitTime - Environment.TickCount` < `TimeUnitMilliseconds` they are going to wait more than expected - The underlying issue is that the timer wake up isn't precise and actually tends to lag behind - solved a similar issue some time ago https://github.com/QuantConnect/Lean/blob/master/Engine/DataFeeds/RealTimeScheduleEventService.cs#L53

Reply:
another issue here is that `DateTime.UtcNow` or similar are only as precise as the system they run on. Windows is notoriously imprecise while *nix systems tend to have sub-millisecond clocks available to C# via dotnet core.